presc/copies/sampling.py

```diff
@@ -296,6 +409,133 @@ def spherical_balancer_sampling(
+    numerical features and with a discrete distribution for the categorical
+    features. The parameters describing the feature space needed to compute the
+    distributions are described in the `feature_parameters` dictionary. Features
+    are assumed to be independent (not correlated).
+
+    Parameters
+    ----------
+    numerical_sampling : function
+        Any of the non balancing numerical sampling functions defined in PRESC:
+        `grid_sampling`, `uniform_sampling`, `normal_sampling`...
+    nsamples : int
+        Number of samples to generate.
+    random_state : int
+        Random seed used to generate the sampling data.
+    feature_parameters : dict of dicts
+        A dictionary with an entry per dataset feature (dictionary keys should
+        be the feature names), where each feature entry must contain a
+        nested dictionary with its categories and their probability. The key for
+        the nested dictionary of categories should be "categories", and the keys
+        for the probabilities should be the category name.
+
+    Returns
+    -------
+    pandas DataFrame
+        Dataset with a generated sampling following the specified numerical
+        sampling distribution for the numerical features and the discrete
+        distribution for the categorical features, following the feature space
+        characterized by the `feature_parameters`.
+    """
+    if random_state is not None:
+        np.random.seed(seed=random_state)
+
+    # Generate the lists of numerical and categorical data
+    features_numerical = []
+    features_categorical = []
+    for feature in feature_parameters:
+        if "categories" in feature_parameters[feature]:
+            features_categorical.append(feature)
+        else:
+            features_numerical.append(feature)
+
+    # Generate feature parameter dictionaries for the numerical/categorical samplers
+    feature_parameters_numerical = {
+        feature: feature_parameters[feature] for feature in features_numerical
+    }
+    feature_parameters_categorical = {
+        feature: feature_parameters[feature] for feature in features_categorical
+    }
+
+    # Generate the numerical/categorical features of each sample separately
+    X_generated_numerical = numerical_sampling(
+        nsamples=nsamples,
+        random_state=random_state,
+        feature_parameters=feature_parameters_numerical,
+    )
+    X_generated_categorical = categorical_sampling(
+        nsamples=nsamples,
+        random_state=random_state,
+        feature_parameters=feature_parameters_categorical,
+    )
+
+    # Combine the numerical/categorical features respecting the structure of the
+    # original dataset
+    X_generated = pd.DataFrame()
+    for feature in feature_parameters:
+        if "categories" in feature_parameters[feature]:
+            X_generated = X_generated.join(
+                X_generated_categorical[[feature]], how="outer"
```

Comment:
Is there a reason to join each column separately? Since you've already got the two `X_generated` DFs, is this different from concatenating them together? Eg.
```python
pd.concat([X_generated_numerical, X_generated_categorical], axis="columns")
```

Reply:
The reason is to maintain the same order of the original columns. With a simple block concatenation the categorical and the numerical columns become separated. I don't know a priori if the model will use the names of the columns (in which case the order wouldn't matter) or if the model will take the same order of the parameters in which it was trained, so I wanted to make sure that the structure of the sampled dataset was exactly the same.

Reply:
As discussed, you can reorder the columns of a DF by passing a list of column names in the desired order, eg. `X[["col2", "col1"]]`.

Reply:
Done.